ksonnet-gen/ksonnet/emit.go

```diff
@@ -0,0 +1,588 @@
+// `ksonnet-lib`, written in Jsonnet.
+func Emit(spec *kubespec.APISpec) ([]byte, error) {
+	root := newRoot(spec)
+
+	m := newMarshaller()
+	return root.emit(m)
+}
+
+//-----------------------------------------------------------------------------
+// Root.
+//-----------------------------------------------------------------------------
+
+// `root` is an abstraction of the root of `k8s.libsonnet`, which can be
+// emitted as Jsonnet code using the `emit` method.
+//
+// `root` contains and manages a set of `groups`, which represent a
+// set of Kubernetes API groups (e.g., core, apps, extensions), and
+// holds all of the logic required to build the `groups` from an
+// `kubespec.APISpec`.
+type root struct {
+	info   *kubespec.SchemaInfo
+	groups groupSet // set of groups, e.g., core, apps, extensions.
+}
+
+func newRoot(spec *kubespec.APISpec) *root {
+	root := root{
+		info:   spec.Info,
+		groups: make(groupSet),
+	}
+
+	for defName, def := range spec.Definitions {
+		parsedName := defName.Parse()
+		root.addDefinition(parsedName, def)
+	}
+
+	return &root
+}
+
+func (root *root) emit(m *marshaller) ([]byte, error) {
+	m.bufferLine("// AUTOGENERATED from the Kubernetes OpenAPI specification. DO NOT MODIFY.")
+	m.bufferLine(fmt.Sprintf("// Kubernetes version: %s", root.info.Version))
+	m.bufferLine(fmt.Sprintf("// SHA of HEAD: %s", getSHARevision()))
+	m.bufferLine("")
+
+	m.bufferLine("{")
+	m.indent()
+
+	// Emit in sorted order so that we can diff the output.
+	for _, group := range root.groups.toSortedSlice() {
+		group.emit(m, root)
+	}
+
+	m.dedent()
+	m.bufferLine("}")
+
+	return m.writeAll()
+}
+
+func (root *root) addDefinition(
+	parsedName *kubespec.ParsedDefinitionName, def *kubespec.SchemaDefinition,
+) {
+	isTopLevel := len(def.TopLevelSpecs) > 0
+	apiObject, err := root.getOrCreateAPIObject(parsedName, isTopLevel)
+	if err != nil {
+		return
+	}
+
+	for propName, prop := range def.Properties {
+		pm := newPropertyMethod(propName, prop, apiObject)
+		apiObject.propertyMethods[propName] = pm
+	}
+}
+
+func (root *root) getOrCreateAPIObject(
+	parsedName *kubespec.ParsedDefinitionName, isTopLevel bool,
+) (*apiObject, error) {
+	if parsedName.Version == nil {
+		return nil, fmt.Errorf(
+			"Can't make API object from name with nil version. package type: '%d' kind: '%s'",
+			parsedName.PackageType,
+			parsedName.Kind)
+	}
+
+	var groupName kubespec.GroupName
+	if parsedName.Group == nil {
+		groupName = "core"
+	} else {
+		groupName = *parsedName.Group
+	}
+
+	group, ok := root.groups[groupName]
+	if !ok {
+		group = newGroup(groupName, root)
+		root.groups[groupName] = group
+	}
+
+	versionedAPI, ok := group.versionedAPIs[*parsedName.Version]
+	if !ok {
+		versionedAPI = newVersionedAPI(*parsedName.Version, group)
+		group.versionedAPIs[*parsedName.Version] = versionedAPI
+	}
+
+	apiObject, ok := versionedAPI.apiObjects[parsedName.Kind]
+	if ok {
+		log.Fatalf("Duplicate object kinds with name '%s'", parsedName.Unparse())
```

Comment:
This doesn't go with the name of this function.  The function is named `getOrCreate` but here we are erroring out if it already exists.  Either have this not be an error or change the name of the method to `createAPIObject`.